doc/design/resource-dependency.md

```diff
@@ -0,0 +1,343 @@
+  In fact, ordering in the first goal can be modeled as data dependency as well except that the receiver doesn't use the data.
+3. **OAM native experience**.
+  OAM has application concepts like Components, Traits, etc.
+  The solution will provide OAM native experience built on top of the generic solution. For example, modeling as traits.
+
+In the following we will first go through the use cases and then propose a solution to satisfy them.
+
+
+## Use Cases
+
+### 1. Web app and database
+
+An applicationConfiguration consists of two components: web and db (database).
+Web component should not be created until db component is created,
+including public endpoint and connection secret to be made ready.
+
+### 2. ML app workflow
+
+An ML applicationConfiguration from [Hypercycle ML](http://www.4paradigm.com/product/hypercycleml)
+has three components:
+
+- preprocess
+- training
+- serving
+
+They have strict ordering:
+
+```
+preprocess -> training -> serving
+```
+
+Each component has to wait for data to be finished processing by the previous stage.
+Today, ML frameworks including Tensorflow do not handle such dependency issue.
+They rely on external tools (e.g. Argo Workflow) to handle it.
+If we want to model them as Components in the same ApplicationConfiguration, the external tools could not help in this case.
+
+### 3. Service-binding Trait
+
+Service binding is an OAM trait that we develop to bind user specified input such as env or file path
+to specific data sources such as secrets or events.
+More background introduction can be found in this [slides](https://docs.google.com/presentation/d/1PseN_8_zZH8clWZJzP8tuRMpz51SxQb-mKeG1f4QVZQ/edit?usp=sharing).
+
+A service binding trait needs to be applied before creating the component.
+Moreover, this kind of trait ordering is quite general and applies to many other traits we have seen.
+
+### 4. NSQ deployment
+
+An [NSQ](https://github.com/nsqio/nsq) cluster is composed with three components: nsqd, nsqlookup and nsqadmin.
+To set up a `nsq` cluster, one needs to first make sure that the `nsqd` component is up, then start and make sure that the `nsqlookup` component is up, and run the `nsqadmin` component at last.
+nsqlookup consumes the result of nsqd component instance, and nsqadmin consumes the result of nsqlookup component instance.
+If we couldn't describe the dependencies order between the components, the component couldn't run properly.
+
+This is similar to [kubernetes#65502](https://github.com/kubernetes/kubernetes/issues/65502) , the issue describes dependencies between containers on the same Pod, while we are discussing dependencies between components.
+
+
+## Proposal
+
+The overall idea is to have each resource object specify data inputs coming from other resources' fields.
+The OAM runtime will build the dependency graph accordingly and only create resources once the data inputs are ready.
+To achieve this, we propose:
+
+1. Add the following `DataInputs` into each resource.
```

Comment:
It is not clear to me what "each resource" refers to here. It seems like this would potentially be surfaced on a `Component` similar to how `parameters` is currently, but it is not present on the `Component` examples below.

Reply:
Both Components and Traits are k8s resources.
It just means the resources that needs data input.

Reply:
TODO: I will update the doc.

Reply:
Thanks :)